### PR TITLE
Closes #1685 Implement data batching for sync messages

### LINKED ISSUE #1685: Batch data sync operations
Given some perf test reports, we should batch replication of `{key, chain}` tuples at the server store level.

### PR DESCRIPTION
**Pull request by albinsuresh**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -97,9 +97,10 @@ subprojects {
   targetCompatibility = 1.6
 
   repositories {
-    if (project.hasProperty('mvnlocal')) {
-      mavenLocal()
-    }
+//    if (project.hasProperty('mvnlocal')) {
+//      mavenLocal()
+//    }
+    mavenLocal()
```
Comment (lorban) [inline build.gradle]: You should drop that change and put `mvnlocal=true` in your `~/.gradle/gradle.properties` file before that gets merged by accident once again.

File: clustered/server/src/main/java/org/ehcache/clustered/server/EhcacheActiveEntity.java [hunk 2]
```diff
@@ -105,6 +105,8 @@
 class EhcacheActiveEntity implements ActiveServerEntity<EhcacheEntityMessage, EhcacheEntityResponse> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(EhcacheActiveEntity.class);
+  static final String SYNC_DATA_SIZE_PROP = "sync.data.size.threshold";
+  private static final long SYNC_DATA_SIZE_THRESHOLD = 64 * 1024;
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/EhcacheActiveEntity.java]: Better prefix that constant with `DEFAULT_`, and I wonder if 64 KB is a reasonable default. I was thinking of something bigger, like 1 MB. Were there perf test results indicating that 64 KB ought to be enough?

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheDataSyncMessage.java [hunk 3]
```diff
@@ -20,17 +20,17 @@
...
-  private final long key;
-  private final Chain chain;
+  private final Map<Long, Chain> chainMap;
 
-  public EhcacheDataSyncMessage(final String cacheId, final long key, final Chain chain) {
+  public EhcacheDataSyncMessage(final String cacheId, final Map<Long, Chain> chainMap) {
     this.cacheId = cacheId;
-    this.key = key;
-    this.chain = chain;
+    this.chainMap = chainMap;
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheDataSyncMessage.java]: Should probably be wrapped in `Collections.unmodifiableMap()` for peace of mind.

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java [hunk 4]
```diff
@@ -101,26 +107,33 @@
...
+        }
+        case DATA: {
           encoder = DATA_SYNC_STRUCT.encoder();
-          EhcacheDataSyncMessage dataSyncMessage = (EhcacheDataSyncMessage) syncMessage;
+          EhcacheDataSyncMessage dataSyncMessage = (EhcacheDataSyncMessage)syncMessage;
           encoder.enm(SYNC_MESSAGE_TYPE_FIELD_NAME, DATA);
           encoder.string(SERVER_STORE_NAME_FIELD, dataSyncMessage.getCacheId());
-          encoder.int64(KEY_FIELD, dataSyncMessage.getKey());
-          encoder.struct(CHAIN_FIELD, (chainEncoder) -> chainCodec.encode(chainEncoder, dataSyncMessage.getChain()));
+          StructArrayEncoder entriesEncoder = encoder.structs(CHAIN_MAP_ENTRIES_SUB_STRUCT);
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java]: Better specify the generic type on `StructArrayEncoder` here.

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java [hunk 5]
```diff
@@ -101,26 +107,33 @@
...
+          for (Map.Entry<Long, Chain> entry : dataSyncMessage.getChainMap().entrySet()) {
+            entriesEncoder.int64(KEY_FIELD, entry.getKey());
+            chainCodec.encode(entriesEncoder.struct(CHAIN_FIELD), entry.getValue());
+            entriesEncoder.next();
+          }
+          entriesEncoder.end();
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java]: You could replace this loop with a call to `encoder.structs(name, lambda)` after setting `StructArrayEncoder`'s generic if you wanted to.

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java [hunk 6]
```diff
@@ -165,14 +178,29 @@ public EhcacheSyncMessage decode(final int concurrencyKey, final byte[] payload)
...
+        return new EhcacheDataSyncMessage(storeName, chainMap);
       default:
         throw new AssertionError("Cannot happen given earlier checks");
     }
   }
 
+  private Map<Long, Chain> decodeChainMapEntries(StructDecoder decoder) {
+    Map<Long, Chain> chainMap = new HashMap<>();
+
+    StructArrayDecoder entriesDecoder = decoder.structs(CHAIN_MAP_ENTRIES_SUB_STRUCT);
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java]: Generic type, please!

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java [hunk 7]
```diff
@@ -165,14 +178,29 @@ public EhcacheSyncMessage decode(final int concurrencyKey, final byte[] payload)
...
+    if (entriesDecoder != null) {
+      for (int i = 0; i < entriesDecoder.length(); i++) {
+        Long key = entriesDecoder.int64(KEY_FIELD);
+        StructDecoder chainDecoder = entriesDecoder.struct(CHAIN_FIELD);
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java]: Generic type, please!

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java [hunk 8]
```diff
@@ -165,14 +178,29 @@ public EhcacheSyncMessage decode(final int concurrencyKey, final byte[] payload)
...
+        Chain chain = chainCodec.decode(chainDecoder);
+        chainMap.put(key, chain);
+        entriesDecoder.next();
+      }
+    }
+    return chainMap;
+  }
+
   private Map<String, ServerStoreConfiguration> decodeStoreConfigurations(StructDecoder decoder) {
```
Comment (lorban) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheSyncMessageCodec.java]: Generic type, please!

File: clustered/server/src/main/java/org/ehcache/clustered/server/EhcacheActiveEntity.java [hunk 9]
```diff
@@ -105,6 +105,8 @@
 class EhcacheActiveEntity implements ActiveServerEntity<EhcacheEntityMessage, EhcacheEntityResponse> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(EhcacheActiveEntity.class);
+  static final String SYNC_DATA_SIZE_PROP = "sync.data.size.threshold";
```
Comment (ljacomet) [inline clustered/server/src/main/java/org/ehcache/clustered/server/EhcacheActiveEntity.java]: Please have `ehcache` appear in there to disambiguate.

File: clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheDataSyncMessage.java (hunk 3 above)
Comment (ljacomet) [inline clustered/server/src/main/java/org/ehcache/clustered/server/internal/messages/EhcacheDataSyncMessage.java]: I would take a copy even, as we do not decide when the encoding of the message happens and we clear the passed map after issuing the send command.

**Comment by albinsuresh** [comment]
Addressed the review comments

**Pull request merged**